juju/juju
# Add new apiserver listener which throttles incoming connections

## Description of change

For models with many agents, it's possible to DDOS the controller (the thundering herd problem) if the controller agent bounces, eg after an upgrade, and everything attempts to re-connect at the same time.

Rate limiting is introduced to logins and general api connections. In addition, when a login is rejected with a retry error, we first pause so that the agent doesn't get to try again too soon.

The connection rate limit is achieved by pausing when accepting api connections. The pause time is based on the connection rate - the higher the rate of incoming connections, the longer the pause. The pause time is increased by 5ms for each new connection received per 10ms, up to a maximum of 5s pause time. 

## QA steps

bootstrap and deploy peer-xplod charm
add units
add new models, rinse and repeat
system should reach steady state without errors
on my lxd system, the units all got to the "reached maximum count 1000" state without noticeable impact on system interactions

## Bug

Partially addresses https://bugs.launchpad.net/juju/+bug/1696113
